### ISSUE TITLE
Dependency shader segfaults with GlslangToSpv

### ISSUE THREAD

**Issue by dchristopherfennell** [KhronosGroup/glslang#2524]
If I have a shader that does lighting and a shader that does fog, glslang::GlslangToSpv will segfault. The issue is a UBO LightBlock that is used in both shaders. The LightBlock has the following:

layout(set = 0, binding = 11) uniform LightBlock
{
    vec4 position;
    vec4 diffuse;
    vec4 ambient;
    vec4 specular;
    vec4 fogvalues;
} lightblock;

The fog shader uses **fogvalues** in the same UBO for exponential fogging. The instance name in the fog shader is **fogblock** instead of **lightblock**.

This compiled, linked, and provided SPV code until commit 478b23295244c9f33eb90a47a99fb3b5cc6e0370 on 11/03/2020 where it now segfaults.

Without going into a debate about using a different UBO for fogvalues, are we no longer able to access the same bind point UBO in more than one shader? Or did I do it incorrectly?

I would like to add that this happens when glslang::SpvOptions disableOptimizer is set to false. If this is set to true, the SPV is generated.

Any help with this will be greatly appreciated.

Thanks,
DC.

**Comment by greg-lunarg** [comment]
Using different instance names for the same named block gives a warning but otherwise appears to be allowed. See the simple example below:

```
$ cat t2a.frag
#version 450 core

layout(set = 0, binding = 11) uniform LightBlock
{
vec4 position;
vec4 diffuse;
vec4 ambient;
vec4 specular;
vec4 fogvalues;
} lightblock;

vec4 getSpec()
{
    return lightblock.specular;
}

$ cat t2b.frag
#version 450 core

layout(set = 0, binding = 11) uniform LightBlock
{
vec4 position;
vec4 diffuse;
vec4 ambient;
vec4 specular;
vec4 fogvalues;
} fogblock;

layout (location=0) out vec4 color;

vec4 getSpec();

void main()
{
     color = getSpec() * fogblock.fogvalues;
}

$ /cygdrive/c/Users/Greg/dev/glslang/build/StandAlone/Release/glslangValidator.exe -V -o t2.frag.spv t2a.frag t2b.frag
t2a.frag
t2b.frag
WARNING: Linking fragment stage: Matched shader interfaces are using different instance names.
    lightblock: "layout( set=0 binding=11 column_major std140) uniform block{layout( column_major std140) uniform highp 4-component vector of float position, layout( column_major std140) uniform highp 4-component vector of float diffuse, layout( column_major std140) uniform highp 4-component vector of float ambient, layout( column_major std140) uniform highp 4-component vector of float specular, layout( column_major std140) uniform highp 4-component vector of float fogvalues}" versus fogblock: "layout( set=0 binding=11 column_major std140) uniform block{layout( column_major std140) uniform highp 4-component vector of float position, layout( column_major std140) uniform highp 4-component vector of float diffuse, layout( column_major std140) uniform highp 4-component vector of float ambient, layout( column_major std140) uniform highp 4-component vector of float specular, layout( column_major std140) uniform highp 4-component vector of float fogvalues}"
```

Certainly the segfault is a legitimate problem with glslang. Can you please send code and command which results in segfault?

@ShchchowAMD

**Comment by ShchchowAMD** [comment]
Looking at it now.

**Comment by dchristopherfennell** [comment]
Hello,

`               options.optimizeSize = true;
                options.disableOptimizer = false;
                options.validate = true;
                options.disassemble = false;

                glslang::GlslangToSpv(*(intermediate), spirv, &logger, &options);
`
This call causes the segfault. If I set **options.disableOptimizer = true**, then I don't segfault.

The different instance names were done because having the same name in the same stage would cause a parsing error some time ago. This no longer happens.

I've attached the fog/light shader in question (modified to use the same instance names per bind point)

[light.frag.txt](https://github.com/KhronosGroup/glslang/files/5944969/light.frag.txt)
[fog.frag.txt](https://github.com/KhronosGroup/glslang/files/5944976/fog.frag.txt)

**Comment by dchristopherfennell** [comment]
Does this help?

Running in debug with **options.disableOptimizer = false**, I get an assert in def_use_manager.cpp line 60/61:
        Instruction* def = GetDef(use_id);
        **assert(def && "Definition is not registered.");**

def is NULL.

**Comment by ShchchowAMD** [comment]
I think the problem happens inside glslang before we converting to SPIR-V. Should be caused by symbol mapping or reflection.
Will update a solution soon.

**Comment by greg-lunarg** [comment]
You didn't supply a main, so I wrote a simple one:

```
#version 450 core
#extension GL_ARB_separate_shader_objects : enable

layout(location = 0) in vec4 iColor;
layout(location = 0) out vec4 oColor;

void FogThis(inout vec4 color);
void LightThis(inout vec4 color);

void main()
{
    vec4 color = iColor;
    FogThis(color);
    LightThis(color);
    oColor = color;
}
```

With yesterday's top of tree for glslang the resulting SPIR-V is valid and runs through yesterday' spirv-opt without problem resulting in valid SPIR-V.

**Comment by greg-lunarg** [comment]
With the main above and yesterday's components do you still see your problem?

**Comment by dchristopherfennell** [comment]
I’ll run this Tuesday with the latest code and let you know. 

Thanks,
DC

**Comment by dchristopherfennell** [comment]
Hello,

I ran it this morning with the latest code without any issues. 
I was able to produce valid SPIR-V.

Thanks for the fix.
DC

**Comment by greg-lunarg** [comment]
Presumed fixed by #2458 or sooner.
